integration/gateway/gateway_test.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright IBM Corp All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package gateway
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"syscall"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/golang/protobuf/proto"
+	"github.com/hyperledger/fabric-protos-go/gateway"
+	"github.com/hyperledger/fabric/integration/nwo"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/tedsuo/ifrit"
+	"google.golang.org/grpc"
+)
+
+var _ = Describe("GatewayService", func() {
+	var (
+		testDir       string
+		network       *nwo.Network
+		org1Peer0     *nwo.Peer
+		conn          *grpc.ClientConn
+		gatewayClient gateway.GatewayClient
+		cancel        context.CancelFunc
+		ctx           context.Context
+		process       ifrit.Process
+	)
+
+	BeforeEach(func() {
+		var err error
+		testDir, err = ioutil.TempDir("", "gateway")
+		Expect(err).NotTo(HaveOccurred())
+
+		client, err := docker.NewClientFromEnv()
+		Expect(err).NotTo(HaveOccurred())
+
+		config := nwo.BasicEtcdRaft()
+		network = nwo.New(config, testDir, client, StartPort(), components)
+
+		network.GatewayEnabled = true
+
+		network.GenerateConfigTree()
+		network.Bootstrap()
+
+		networkRunner := network.NetworkGroupRunner()
+		process = ifrit.Invoke(networkRunner)
+		Eventually(process.Ready(), network.EventuallyTimeout).Should(BeClosed())
+
+		orderer := network.Orderer("orderer")
+		network.CreateAndJoinChannel(orderer, "testchannel")
+		network.UpdateChannelAnchors(orderer, "testchannel")
+		network.VerifyMembership(
+			network.PeersWithChannel("testchannel"),
+			"testchannel",
+		)
+		nwo.EnableCapabilities(
+			network,
+			"testchannel",
+			"Application", "V2_0",
+			orderer,
+			network.PeersWithChannel("testchannel")...,
+		)
+
+		chaincode := nwo.Chaincode{
+			Name:            "gatewaycc",
+			Version:         "0.0",
+			Path:            components.Build("github.com/hyperledger/fabric/integration/chaincode/simple/cmd"),
+			Lang:            "binary",
+			PackageFile:     filepath.Join(testDir, "gatewaycc.tar.gz"),
+			Ctor:            `{"Args":[]}`,
+			SignaturePolicy: `AND ('Org1MSP.peer')`,
+			Sequence:        "1",
+			InitRequired:    false,
+			Label:           "gatewaycc_label",
+		}
+
+		nwo.DeployChaincode(network, "testchannel", orderer, chaincode)
+
+		org1Peer0 = network.Peer("Org1", "peer0")
+		conn = network.PeerClientConn(org1Peer0)
+		gatewayClient = gateway.NewGatewayClient(conn)
+		ctx, cancel = context.WithTimeout(context.Background(), network.EventuallyTimeout)
```

Comment:
As was done in the commit I pointed you to, please move these to the `It` blocks (tests) and use a `defer` to handle the cancel and cleanup instead of creating a block level `var` and relying on `AfterEach` to clean this bit up.

After that change, I'm happy with what's here.

Reply:
No problem- moved. Out of interest why wasn't it a good idea to remove the duplication and rely on `AfterEach`?

Reply:
I started writing a reply but it looks like I lost it so I'll provide the highlights.

The first is that it's a common idiom in Go to acquire a connection and have the `defer conn.Close()` next to it after the error check. When it's missing, the reader then goes searching for where it's done. There's a similar rationale (and generally more important) around calling the cancel function in a defer.

The second is philosophical. There's a general preference for DRY production code and DAMP (and/or WET) test code. This makes it easy for people to jump to a failing test and quickly get context without having to walk through all of the callbacks used to setup for the test.